sources/academy/webscraping/scraping_basics_python/10_crawling.md

```diff
@@ -1,15 +1,305 @@
+    data.append(item)
+
+with open("products.csv", "w") as file:
+    export_csv(file, data)
+
+with open("products.json", "w") as file:
+    export_json(file, data)
+```
+
+## Extracting vendor name
+
+Each product URL points to a so-called _product detail page_, or PDP. If we open one of the product URLs in the browser, e.g. the one about [Sony XBR-950G BRAVIA](https://warehouse-theme-metal.myshopify.com/products/sony-xbr-65x950g-65-class-64-5-diag-bravia-4k-hdr-ultra-hd-tv), we can see that it contains a vendor name, [SKU](https://en.wikipedia.org/wiki/Stock_keeping_unit), number of reviews, product images, product variants, stock availability, description, and perhaps more.
+
+![Product detail page](./images/pdp.png)
+
+Depending on what's valuable for our use case, we can now use the same techniques as in previous lessons to extract any of the above. As a demonstration, let's scrape the vendor name. In browser DevTools, we can see that the HTML around the vendor name has the following structure:
+
+```html
+<div class="product-meta">
+  <h1 class="product-meta__title heading h1">
+    Sony XBR-950G BRAVIA 4K HDR Ultra HD TV
+  </h1>
+  <div class="product-meta__label-list">
+    ...
+  </div>
+  <div class="product-meta__reference">
+    <!-- highlight-next-line -->
+    <a class="product-meta__vendor link link--accented" href="/collections/sony">
+        <!-- highlight-next-line -->
+        Sony
+    <!-- highlight-next-line -->
+    </a>
+    <span class="product-meta__sku">
+      SKU:
+      <span class="product-meta__sku-number">SON-985594-XBR-65</span>
+    </span>
+  </div>
+  <a href="#product-reviews" class="product-meta__reviews-badge link" data-offset="30">
+    <div class="rating">
+      <div class="rating__stars" role="img" aria-label="4.0 out of 5.0 stars">
+        ...
+      </div>
+      <span class="rating__caption">3 reviews</span>
+    </div>
+  </a>
+  ...
+</div>
+```
+
+It looks like using a CSS selector to locate the element with the `product-meta__vendor` class, and then extracting its text, should be enough to get the vendor name as a string:
+
+```py
+vendor = product_soup.select_one(".product-meta__vendor").text.strip()
+```
+
+But where do we put this line in our program?
+
+## Crawling product detail pages
+
+In the `data` loop we're already going through all the products. Let's expand it to include downloading the product detail page, parsing it, extracting the vendor's name, and adding it as a new key in the item's dictionary:
+
+```py
+...
+
+listing_url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+listing_soup = download(listing_url)
+
+data = []
+for product in listing_soup.select(".product-item"):
+    item = parse_product(product, listing_url)
+    # highlight-next-line
+    product_soup = download(item["url"])
+    # highlight-next-line
+    item["vendor"] = product_soup.select_one(".product-meta__vendor").text.strip()
+    data.append(item)
+
+...
+```
```

Comment:
It runs surprisingly long:

```text
$ time python review_pr_1244.py 

real    0m6,548s
user    0m1,917s
sys     0m0,070s
```

for scraping ~20 items. But I believe it's the website, right?

Reply:
The website is actually quite fast. At least for me 😅

Reply:
What about merging this and creating an issue for the remaining problems so that it can go on-line and start getting indexed, used and so on?

Reply:
Once it's in the git history, the repo gets bloated forever. I'm for optimizing the images first, unless me pushing to this branch didn't already do the damage.

Reply:
I'd even do it off my stint, but unfortunately my family situation is just getting worse, so I barely work on anything. OTOH anyone who's able to run CLI commands (e.g. [this](https://github.com/343dev/optimizt)) and push to branches should be able to do it.

Reply:
I didn't measure the speed of the script, but yes, it's slow and it's kinda expected, it's just a loop with requests. It's gonna be one of the reasons why to employ Crawlee in the next lesson.

I see the size of the images as a bigger problem. Once it's in the git history, the repo gets bloated forever. I'm for optimizing the images first, unless me pushing to this branch didn't already do the damage.

I'd even do it off my stint, but unfortunately my family situation is just getting worse, so I barely work on anything. OTOH anyone who's able to run CLI commands (e.g. [this](https://github.com/343dev/optimizt)) and push to branches should be able to do it.

Reply:
```
+243 KB sources/academy/webscraping/scraping_basics_python/images/pdp.png
+494 KB sources/academy/webscraping/scraping_basics_python/images/variants.png
```